File: generated_code/python/solution_124_1.py
import sys
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        # Initialize the global maximum path sum to a very small integer
        self.max_sum = -sys.maxsize
        
        def get_max_gain(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            
            # Recursively get the maximum path sum from left and right subtrees.
            # If a subtree path sum is negative, we ignore it by taking max(..., 0).
            left_gain = max(get_max_gain(node.left), 0)
            right_gain = max(get_max_gain(node.right), 0)
            
            # Price of the new path would be the parent node value plus the contributions from both subtrees
            current_path_sum = node.val + left_gain + right_gain
            
            # Update the global maximum path sum if the current path is better
            self.max_sum = max(self.max_sum, current_path_sum)
            
            # For the parent node call, we can only choose one child's path to maintain a valid tree path
            return node.val + max(left_gain, right_gain)
        
        get_max_gain(root)
        return self.max_sum

File: generated_code/cpp/solution_124_1.cpp
#include <algorithm>
#include <climits>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int max_sum;

    // Helper function that returns the maximum path sum starting from the current node
    // and extending down into one of its subtrees.
    int calculateMaxGain(TreeNode* node) {
        if (!node) {
            return 0;
        }

        // Recursively find the maximum path sum of the left and right subtrees.
        // If a subtree returns a negative sum, we choose not to include it (hence std::max(..., 0)).
        int left_gain = std::max(calculateMaxGain(node->left), 0);
        int right_gain = std::max(calculateMaxGain(node->right), 0);

        // Price of a new path where the current node is the highest node (the "turn" point).
        int current_path_sum = node->val + left_gain + right_gain;

        // Update the global maximum path sum found so far.
        max_sum = std::max(max_sum, current_path_sum);

        // For the parent call, we can only choose one child path to maintain a valid path.
        return node->val + std::max(left_gain, right_gain);
    }

public:
    int maxPathSum(TreeNode* root) {
        max_sum = INT_MIN;
        calculateMaxGain(root);
        return max_sum;
    }
};